From 20839d6109ae6d07a84b6fd56660d7b59ce9056c Mon Sep 17 00:00:00 2001
From: MeFisto94 <MeFisto94@users.noreply.github.com>
Date: Fri, 20 Aug 2021 00:09:35 +0200
Subject: [PATCH] Added an overload of Semih_Network#TryToJoinGame that takes
 an IP-Port Pair and added calls to the command line parsing for connect and
 server hosting

---
 Semih_Network.cs | 64 +++++++++++++++++++++++++++++++++++++++++++++-----------
 1 file changed, 52 insertions(+), 12 deletions(-)

diff --git a/Semih_Network.cs b/Semih_Network.cs
index 3dc8957..54c6001 100644
--- a/Semih_Network.cs
+++ b/Semih_Network.cs
@@ -632,7 +632,7 @@ public class Semih_Network : MonoBehaviour, SceneEventInterface
 				{
 					Semih_Network.OnJoinResult(psteamIDRemote, message_InitiateConnectionResult.result);
 				}
-				
+
 				// Sharkkit Dedicated - Override this check because when connecting to a dedicated server, the steamId is not known before the handshake.
 				// Sharkkit Dedicated - In Vanilla it also feels rare that a client gets an InitiateResult request from a server they didn't connect to, anyway...
 				//if (awaitingConnectionResponseUser.IsValid() && awaitingConnectionResponseUser == psteamIDRemote)
@@ -926,7 +926,13 @@ public class Semih_Network : MonoBehaviour, SceneEventInterface
 		}
 	}
 
-	public void TryToJoinGame(CSteamID hostID, string password)
+	// Sharkkit Dedicated Start - Move method body down into a method also accepting ip:port addresses. 
+	public void TryToJoinGame(CSteamID hostId, string password)
+	{
+		TryToJoinGame(hostId, password, null, 0);
+	}
+
+	public void TryToJoinGame(CSteamID hostId, string password, string ipAddress, ushort port)
 	{
 		if (IsConnectedToHost)
 		{
@@ -934,15 +940,27 @@ public class Semih_Network : MonoBehaviour, SceneEventInterface
 		}
 		else if (!awaitingConnectionResponseUser.IsValid())
 		{
-			this.hostID = hostID;
-			awaitingConnectionResponseUser = hostID;
+			this.hostID = hostId;
+			awaitingConnectionResponseUser = hostId;
 			isHost = false;
 			gameManagerStartCalled = false;
 			worldLoadedCoroutine = null;
-			Helper.LogBuild("Sending request to join " + SteamFriends.GetFriendPersonaName(hostID) + "'s game");
-			SendP2P(hostID, new Message_InitiateConnection(Messages.InitiateSession, Settings.AppBuildID, password), EP2PSend.k_EP2PSendReliable, NetworkChannel.Channel_Session);
+			
+			if (ipAddress != null)
+			{
+				// Sharkkit Dedicated - The Network Wrapper needs to prepare a connection
+				NetworkWrapper.ConnectToIP(hostId, ipAddress, port);
+				Helper.LogBuild($"Sending request to join the dedicated server at {ipAddress}:{port}");
+			}
+			else
+			{
+				Helper.LogBuild("Sending request to join " + SteamFriends.GetFriendPersonaName(hostId) + "'s game");
+			}
+
+			SendP2P(hostId, new Message_InitiateConnection(Messages.InitiateSession, Settings.AppBuildID, password), EP2PSend.k_EP2PSendReliable, NetworkChannel.Channel_Session);
 		}
 	}
+	// Sharkkit Dedicated End - Move method body down into a method also accepting ip:port addresses.
 
 	public void LeaveGame(DisconnectReason reason, SceneName sceneName, bool saveGame)
 	{
@@ -980,8 +998,7 @@ public class Semih_Network : MonoBehaviour, SceneEventInterface
 	{
 		if (awaitingConnectionResponseUser.IsValid())
 		{
-			// Sharkkit Dedicated - Redirect call to Network Wrapper
-			NetworkWrapper.CloseP2PSessionWithUser(HostID);
+			NetworkWrapper.CloseP2PSessionWithUser(HostID); // Sharkkit Dedicated - Redirect call to Network Wrapper
 			awaitingConnectionResponseUser.Clear();
 			hostID.Clear();
 		}
@@ -1315,13 +1332,36 @@ public class Semih_Network : MonoBehaviour, SceneEventInterface
 		{
 			return;
 		}
-		foreach (string text2 in commandLineArgs)
+		
+		// Sharkkit Dedicated Start - Parse the dedicated commandLine and host a server
+		if (CommandLine.ParseDedicated(Array.AsReadOnly(commandLineArgs), out var listenArguments))
+		{
+			// TODO: Delay call to ListenOn until after a world has been loaded, to prevent crashes when connecting too early
+			NetworkWrapper.ListenOn(listenArguments);
+		}
+		// Sharkkit Dedicated End - Parse the dedicated commandLine and host a server
+		
+		foreach (var arg in commandLineArgs)
 		{
-			text += text2;
-			if (text2.Contains(CONNECT))
+			text += arg;
+			if (arg.Contains(CONNECT))
 			{
-				HandleConnectArgument(text2);
+				HandleConnectArgument(arg);
+			}
+			// Sharkkit Dedicated Start - Parse ips from the commandline and auto-connect
+			else if (CommandLine.ParseIpConnect(arg, out var addr, out var port, out var steamId))
+			{
+				// The following is a copy of the original join code, but calling the correct TryToJoinGame Method
+				var startMenuScreen = FindObjectOfType<StartMenuScreen>();
+				// TODO: Password
+				TryToJoinGame(steamId, string.Empty, addr, port);
+				if (startMenuScreen != null && startMenuScreen.joinGameBox != null)
+				{
+					startMenuScreen.joinGameBox.connectingBox.gameObject.SetActive(true);
+					startMenuScreen.joinGameBox.connectingBox.StartConnectTimeoutTimer();
+				}
 			}
+			// Sharkkit Dedicated End - Parse ips from the commandline and auto-connect
 		}
 		string[] array = text.Replace(" ", "").ToLower().Split('-');
 		foreach (string text3 in array)
-- 
2.7.4

